cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogAppenderInitializer.java

```diff
@@ -36,47 +40,54 @@
   private final LogAppender logAppender;
+  private final String rootLoggerName = Logger.ROOT_LOGGER_NAME;
 
   @Inject
   public LogAppenderInitializer(LogAppender logAppender) {
     this.logAppender = logAppender;
   }
 
   public void initialize() {
-    if (initMap.putIfAbsent(org.slf4j.Logger.ROOT_LOGGER_NAME, logAppender.getName()) != null) {
+    if (initMap.putIfAbsent(Logger.ROOT_LOGGER_NAME, logAppender.getName()) != null) {
       // Already initialized.
       LOG.warn("Log appender {} is already initialized.", logAppender.getName());
       return;
     }
-
-    initialize(org.slf4j.Logger.ROOT_LOGGER_NAME);
+    initialize(rootLoggerName);
   }
 
   @VisibleForTesting
-  public void initialize(String rootLoggerName) {
-    ILoggerFactory loggerFactory = LoggerFactory.getILoggerFactory();
-    // TODO: fix logging issue in mapreduce:  ENG-3279
-    if (!(loggerFactory instanceof LoggerContext)) {
-      LOG.warn("LoggerFactory is not a logback LoggerContext. No log appender is added. " +
-                 "Logback might not be in the classpath");
-      return;
-    }
-
-    LoggerContext loggerContext = (LoggerContext) loggerFactory;
-    Logger rootLogger = loggerContext.getLogger(rootLoggerName);
+  public void initialize(String loggerName) {
+    LoggerContext loggerContext = getLoggerContext();
+    if (loggerContext != null) {
+      Logger rootLogger = loggerContext.getLogger(loggerName);
+      
+      LOG.info("Initializing log appender {}", logAppender.getName());
 
-    LOG.info("Initializing log appender {}", logAppender.getName());
+      // Display any errors during initialization of log appender to console
+      StatusManager statusManager = loggerContext.getStatusManager();
+      OnConsoleStatusListener onConsoleListener = new OnConsoleStatusListener();
+      statusManager.add(onConsoleListener);
 
+      logAppender.setContext(loggerContext);
+      logAppender.start();
 
-    // Display any errors during initialization of log appender to console
-    StatusManager statusManager = loggerContext.getStatusManager();
-    OnConsoleStatusListener onConsoleListener = new OnConsoleStatusListener();
-    statusManager.add(onConsoleListener);
-
-    logAppender.setContext(loggerContext);
-    logAppender.start();
+      rootLogger.addAppender(logAppender);
+    }
+  }
 
-    rootLogger.addAppender(logAppender);
+  public void setLogLevels(Map<String, String> logPairs) {
+    LoggerContext loggerContext = getLoggerContext();
+    if (loggerContext != null) {
+      for (Map.Entry<String, String> entry : logPairs.entrySet()) {
+        String loggerName = entry.getKey();
+        String logLevel = entry.getValue();
+        Logger logger = loggerContext.getLogger(loggerName);
+        LOG.info("Log level of {} changed from {} to {}", loggerName, logger.getLevel(), logLevel);
+        logger.setLevel(Level.toLevel(logLevel));
+        logger.addAppender(logAppender);
```

Comment:
Why need to add appender when changing log level? Will the log get logged twice?